Issue #2246 release notes agent notification templates

GrKolychev opened the pull request:
This PR relates to issue #2246 and brings iplemetation of template notification service to build automatically emails according to templates.

GrKolychev committed:
Issue #2246 created mail templates for notification agent

GrKolychev committed:
Merge branch 'issue_2242_release_notes_agent' into Issue_#2246_release_notes_agent_notification_templates

GrKolychev committed:
Issue #2246 filled application.properties

GrKolychev committed:
Issue #2246 fixed injection

GrKolychev committed:
Issue #2246 Added license, fixed html style, remove unnecessary dependencies

GrKolychev committed:
Issue #2246 added url field to GitHubIssue entity

GrKolychev committed:
Issue #2246 fixed pmd issue

GrKolychev committed:
Issue #2246 added licenses

GrKolychev committed:
Merge branch 'issue_2242_release_notes_agent' into Issue_#2246_release_notes_agent_notification_templates

# Conflicts:
#	release-notes-agent/build.gradle
#	release-notes-agent/src/main/java/com/epam/release/notes/agent/cli/ReleaseNotesNotifier.java
#	release-notes-agent/src/main/java/com/epam/release/notes/agent/service/action/mail/TemplateNotificationServiceImpl.java
#	release-notes-agent/src/main/resources/application.properties

GrKolychev committed:
Issue #2246 changed template resolver to FileTemplateResolver

GrKolychev committed:
Issue #2246 changed gradle.properties

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/configuration/MailTemplateConfig.java:
```diff
@@ -0,0 +1,59 @@
+import org.thymeleaf.templateresolver.FileTemplateResolver;
+
+@Configuration
+public class MailTemplateConfig {
```
better not to call it `mail` because we can use it not only for mail.
I believe `ThymeleafTemplateConfig` would be fine or you can provide your variant

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/configuration/MailTemplateConfig.java:
```diff
@@ -0,0 +1,59 @@
+public class MailTemplateConfig {
+
+    private static final String EMAIL_TEMPLATE_ENCODING = "UTF-8";
+    private static final String PERIOD = ".";
```
I'm not sure that I catch name idea

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/entity/github/GitHubIssue.java:
```diff
@@ -16,15 +16,18 @@
     Long number;
     String title;
     String body;
+    String html_url;
```
please use java convention -> `htmlUrl`

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/service/action/mail/TemplateNotificationServiceImpl.java:
```diff
@@ -17,15 +17,64 @@
 
 @Service
-public class TemplateNotificationServiceImpl implements TemplateNotificationService {
+public class TemplateNotificationServiceImpl implements TemplateNotificationService{
```
`TemplateNotificationService{` -> `TemplateNotificationService {`

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/service/action/mail/TemplateNotificationServiceImpl.java:
```diff
@@ -17,15 +17,64 @@
+    private final String emailToAdminTitle;
+    private final String emailToSubscribersTitle;
+
+    public TemplateNotificationServiceImpl(
```
since this implementation specifically about `email` let's name it `EmailTemplateNotificationService`

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/service/action/mail/TemplateNotificationServiceImpl.java:
```diff
@@ -17,15 +17,64 @@
+                                 final List<JiraIssue> jiraIssues,
+                                 final List<GitHubIssue> gitHubIssues) {
+        final Context context = new Context(Locale.US);
+        context.setVariable("oldVersion", oldVersion);
```
all sting literals should be moved to constants

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/service/action/mail/TemplateNotificationServiceImpl.java:
```diff
@@ -17,15 +17,64 @@
+        final Context context = new Context(Locale.US);
+        context.setVariable("oldVersion", oldVersion);
+        context.setVariable("newVersion", newVersion);
+        if (validateIssuesParameters(jiraIssues, gitHubIssues)) {
```
maybe we need to rename this method some how or add a comment about flow of this method to be more specific and clear when we send email about a changes and when we will send email to admin

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/service/action/mail/TemplateNotificationServiceImpl.java:
```diff
@@ -17,15 +17,64 @@
+                    .body(htmlEmailContent)
+                    .build();
+        }
+        final String htmlEmailContent = templateEngine.process(emailToAdminTemplateName, context);
```
I am not sure that we need to send email to admin in all cases except when we have issues, lets discuss it internally

GrKolychev committed:
Issue #2246
- fixed review issues
- changed method populate() in EmailTemplateNotificationService

GrKolychev committed:
Issue #2246
- fixed review issues
- changed method populate() in EmailTemplateNotificationService

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/service/action/mail/EmailTemplateNotificationService.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    @Override
+    public EmailContent populate(final String oldVersion,
```
why you didn't change this method signature as discussed? by moving all parameters to separate object? is there any problem with that?

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/configuration/TymeleafTemplateConfig.java:
```diff
@@ -0,0 +1,59 @@
+    @Bean
+    public AbstractConfigurableTemplateResolver fileTemplateResolver() {
+        final AbstractConfigurableTemplateResolver templateResolver = new FileTemplateResolver();
+        templateResolver.setPrefix(resolverPrefix);
```
is this resolver can work with files from classpath? what should I use to say to resolver that this file from resources?

GrKolychev committed:
Issue #2246
- writing test

GrKolychev committed:
Merge branch 'issue_2242_release_notes_agent' into Issue_#2246_release_notes_agent_notification_templates

# Conflicts:
#	release-notes-agent/src/main/java/com/epam/release/notes/agent/service/action/mail/TemplateNotificationServiceImpl.java
#	release-notes-agent/src/main/java/com/epam/release/notes/agent/service/jira/JiraIssueServiceImpl.java

GrKolychev committed:
Issue #2246
- Changed approach to populate email content

GrKolychev committed:
Issue #2246
- Create test to EmailTemplateNotificationService

GrKolychev committed:
Issue #2246
- Create test to EmailTemplateNotificationService

GrKolychev committed:
Merge remote-tracking branch 'origin/Issue_#2246_release_notes_agent_notification_templates' into Issue_#2246_release_notes_agent_notification_templates

# Conflicts:
#	release-notes-agent/src/test/java/com/epam/release/notes/agent/service/action/mail/EmailTemplateNotificationServiceTest.java

GrKolychev committed:
Issue #2246
-added conditions to thymleaf template
-changed populate() method

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/cli/ReleaseIssuesGrabber.java:
```diff
@@ -19,22 +19,20 @@
 
 @Slf4j
 @ShellComponent
+@AllArgsConstructor
```
lets remove this class at all

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/configuration/ThymeleafTemplateConfig.java:
```diff
@@ -0,0 +1,69 @@
+            templateResolver = new FileTemplateResolver();
+            templateResolver.setPrefix(fileResolverPrefix);
+        }
+        templateResolver.setPrefix(fileResolverPrefix);
```
this line should be deleted because it will possibly rewrite CLASS_LOADER_RESOLVER_PREFIX

SilinPavel commented on release-notes-agent/src/main/java/com/epam/release/notes/agent/service/action/mail/EmailTemplateNotificationService.java:
```diff
@@ -0,0 +1,93 @@
+    private final String emailToAdminTitle;
+    private final String emailToSubscribersTitle;
+
+    public EmailTemplateNotificationService(
```
this class also should decide to whom this email will be sent, see how `EmailContent` class is configured, there is `recipients` field

SilinPavel replied:
also please be accurate with null for recipients, pease fill it with empty list of null value was provided (see ListUtils.emptyIfNull functionality of Apache common lib)

SilinPavel replied:
to be able to inject such `@Value` as `List<String>` use this notation of  Spring Expression Lang
`@Value("#{'${release.notes.agent.subscriber.emails}'.split(',')}") List<String subscriberEmails`

SilinPavel commented on release-notes-agent/src/main/resources/application.properties:
```diff
@@ -21,8 +21,16 @@ jira.auth.token=${JIRA_TOKEN}
 release.notes.agent.admin.email=
 release.notes.agent.subscribers=
+release.notes.agent.path.to.system.folder.with.templates=
+release.notes.agent.classpath.to.templates=
```
as we discussed it could be removed and replaced with constant value

SilinPavel commented on release-notes-agent/src/main/resources/application.properties:
```diff
@@ -21,8 +21,16 @@ jira.auth.token=${JIRA_TOKEN}
 release.notes.agent.subscribers=
+release.notes.agent.path.to.system.folder.with.templates=
+release.notes.agent.classpath.to.templates=
+release.notes.agent.type.of.template.files=
```
do we actually need this at all? can we store html template in files without any extension? I think yes, since after all we just fill it and put to email body

GrKolychev committed:
Issue #2246
- fixed review issues
